### ISSUE TITLE
Why there is a sharp decrease of PPL between two epochs of the transformer model? 

### ISSUE THREAD

**Issue by gushu333** [pytorch/fairseq#378]
Hi.
I found a strange phenomenon during training with the transformer model.
According to the output log, I found that there was a sharp decrease of PPL between two epochs:
***
{"epoch": 1, "update": 0.9892654178067776, "loss": "7.227", "nll_loss": "5.981", "ppl": "63.16", "wps": 43621, "ups     ": "1.5", "wpb": 28970, "bsz": 947, "num_updates": 4701, "lr": 0.0009224330820691299, "gnorm": "0.851", "clip": "0%     ", "oom": 0.0, "wall": 3154, "train_wall": 2983}
 112 {"epoch": 1, "update": 0.9997895179962113, "loss": "7.207", "nll_loss": "5.958", "ppl": "62.17", "wps": 43629, "ups     ": "1.5", "wpb": 28968, "bsz": 947, "num_updates": 4751, "lr": 0.0009175663546331154, "gnorm": "0.847", "clip": "0%     ", "oom": 0.0, "wall": 3187, "train_wall": 3014}
 113 {"epoch": 1, "loss": "7.207", "nll_loss": "5.958", "ppl": "62.17", "wps": 43629, "ups": "1.5", "wpb": 28968, "bsz":      947, "num_updates": 4751, "lr": 0.0009175663546331154, "gnorm": "0.847", "clip": "0%", "oom": 0.0, "wall": 3187, "     train_wall": 3014}
 114 {"epoch": 1, "valid_loss": 4.9541451967353405, "valid_nll_loss": 3.3313119157728357, "valid_ppl": "10.07", "num_upd     ates": 4751, "best": 4.9541451967353405}
 115 {"epoch": 2, "update": 1.0105241001894338, "loss": "5.260", "nll_loss": "3.732", "ppl": "13.29", "wps": 44956, "ups     ": "0.9", "wpb": 29004, "bsz": 978, "num_updates": 4802, "lr": 0.0009126808071427712, "gnorm": "0.486", "clip": "0%     ", "oom": 0.0, "wall": 3244, "train_wall": 3046}
 116 {"epoch": 2, "update": 1.0210482003788677, "loss": "5.285", "nll_loss": "3.761", "ppl": "13.56", "wps": 44880, "ups     ": "1.1", "wpb": 29142, "bsz": 964, "num_updates": 4852, "lr": 0.0009079660281786226, "gnorm": "0.482", "clip": "0%     ", "oom": 0.0, "wall": 3277, "train_wall": 3078}
 117 {"epoch": 2, "update": 1.0315723005683015, "loss": "5.268", "nll_loss": "3.741", "ppl": "13.37", "wps": 44665, "ups     ": "1.2", "wpb": 29093, "bsz": 941, "num_updates": 4902, "lr": 0.0009033235699638438, "gnorm": "0.488", "clip": "0%     ", "oom": 0.0, "wall": 3310, "train_wall": 3109}
***
The PPL falls from about 60 to 13. The data I used is the WMT14 en-de data, and both the base and big models have the similar phenomenon.
So, Is this normal?  What the possible reasons for this?
Thank you !

**Comment by maxdel** [comment]
Hello,

I believe it is normal and the reason is that the ppl metric you see during training is averaged across all batches in the epoch.

So the first batches (were the model is weaker) of the epoch 1 contribute negatively to the average loss you see. 

However,  when the epoch ends, the ppl metric counter resets and you start to see much better performance from the first batch. 

Hope it helps.

**Comment by YDspeed** [comment]
hello, I have some questions about output log .
what is  the meaning of 'wps'、'wpb'、'ups'、‘wall’ and 'train wall' 
Thank you
